File: Samples/AWFunctional/Mapping/ProductModelIllustrationMap.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.Data.Entity.ModelConfiguration;

namespace AdventureWorksModel
{
    public class ProductModelIllustrationMap : EntityTypeConfiguration<ProductModelIllustration>
    {
        public ProductModelIllustrationMap()
        {
            // Primary Key
            HasKey(t => new { t.ProductModelID, t.IllustrationID });

            // Properties
            Property(t => t.ProductModelID)
                .HasDatabaseGeneratedOption(DatabaseGeneratedOption.None);

            Property(t => t.IllustrationID)
                .HasDatabaseGeneratedOption(DatabaseGeneratedOption.None);

            // Table & Column Mappings
            ToTable("ProductModelIllustration", "Production");
            Property(t => t.ProductModelID).HasColumnName("ProductModelID");
            Property(t => t.IllustrationID).HasColumnName("IllustrationID");
            Property(t => t.ModifiedDate).HasColumnName("ModifiedDate").IsConcurrencyToken(false);

            // Relationships
            HasRequired(t => t.Illustration)
                .WithMany(t => t.ProductModelIllustration)
                .HasForeignKey(d => d.IllustrationID);
            HasRequired(t => t.ProductModel)
                .WithMany(t => t.ProductModelIllustration)
                .HasForeignKey(d => d.ProductModelID);

        }
    }
}

File: Test/AdventureWorksLegacyModelVB/Mapping/ProductModelIllustrationMap.vb
Imports System.ComponentModel.DataAnnotations.Schema
Imports System.Data.Entity.ModelConfiguration

Imports Microsoft.EntityFrameworkCore
Imports Microsoft.EntityFrameworkCore.Metadata.Builders

Namespace AW.Mapping
	Public Class ProductModelIllustrationMap
		Inherits EntityTypeConfiguration(Of ProductModelIllustration)

		Public Sub New()
			' Primary Key
			HasKey(Function(t) New With {
				Key t.ProductModelID,
				Key t.IllustrationID
			})

			' Properties
			[Property](Function(t) t.ProductModelID).HasDatabaseGeneratedOption(DatabaseGeneratedOption.None)

			[Property](Function(t) t.IllustrationID).HasDatabaseGeneratedOption(DatabaseGeneratedOption.None)

			' Table & Column Mappings
			ToTable("ProductModelIllustration", "Production")
			[Property](Function(t) t.ProductModelID).HasColumnName("ProductModelID")
			[Property](Function(t) t.IllustrationID).HasColumnName("IllustrationID")
			[Property](Function(t) t.ModifiedDate).HasColumnName("ModifiedDate") '.IsConcurrencyToken();

			' Relationships
			HasRequired(Function(t) t.Illustration).WithMany(Function(t) t.ProductModelIllustration).HasForeignKey(Function(d) d.IllustrationID)
			HasRequired(Function(t) t.ProductModel).WithMany(Function(t) t.ProductModelIllustration).HasForeignKey(Function(d) d.ProductModelID)
		End Sub
	End Class

	Public Module Mapper
		<System.Runtime.CompilerServices.Extension> _
		Public Sub Map(ByVal builder As EntityTypeBuilder(Of ProductModelIllustration))
			builder.HasKey(Function(t) New With {
				Key t.ProductModelID,
				Key t.IllustrationID
			})

			' Properties
			builder.Property(Function(t) t.ProductModelID).ValueGeneratedNever()

			builder.Property(Function(t) t.IllustrationID).ValueGeneratedNever()

			' Table & Column Mappings
			builder.ToTable("ProductModelIllustration", "Production")
			builder.Property(Function(t) t.ProductModelID).HasColumnName("ProductModelID")
			builder.Property(Function(t) t.IllustrationID).HasColumnName("IllustrationID")
			builder.Property(Function(t) t.ModifiedDate).HasColumnName("ModifiedDate") '.IsConcurrencyToken();

			' Relationships
			builder.HasOne(Function(t) t.Illustration).WithMany(Function(t) t.ProductModelIllustration).HasForeignKey(Function(d) d.IllustrationID)
			builder.HasOne(Function(t) t.ProductModel).WithMany(Function(t) t.ProductModelIllustration).HasForeignKey(Function(d) d.ProductModelID)
		End Sub
	End Module
End Namespace